basic/airdrop-with-withdraw/src/lib.rs

```diff
@@ -0,0 +1,83 @@
+use scrypto::prelude::*;
+
+
+blueprint! {
+
+    struct AirdropWithWithdraw {
+        admin_badge: Address,
+        recipients: HashMap<Address, Decimal>,
+        tokens : Vault
+    }
+
+    impl AirdropWithWithdraw {
+        
+        pub fn new(token : Address) -> (Component, Bucket) {
+
+            let admin_badge = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+                                .initial_supply_fungible(1);
+
+
+            
+            let component = Self {
+                admin_badge : admin_badge.resource_address(),
+                recipients :  HashMap::new(),
+                tokens : Vault::new(token),
+            }
+            .instantiate();
+
+            return (component, admin_badge)
+        }
+
+        fn get_required_tokens_internal(&self) -> Decimal {
+            let mut required_tokens  = Decimal::zero();
+            for val in self.recipients.values() {
+                required_tokens = required_tokens + *val ;  
+            }
+            return required_tokens;
+        }
+
+        #[auth(admin_badge)]
+        pub fn get_required_tokens(&self) -> Decimal {
+            return self.get_required_tokens_internal();
+        }
+
+        #[auth(admin_badge)]
+        pub fn add_recipient(&mut self, recipient: Address, tokens : Decimal)  {
```

Comment:
Using the recipient address is not the resource oriented way. I suggest using badges. In the `new` function, create a new token for the badge:
```
// Used to mint new user badges
let minter_badge = ResourceBuilder::new_fungible(DIVISIBILITY_MINIMUM)
                                   .initial_supply_fungible(1); // Store this in a vault on the component

let user_badge_def = ResourceBuilder::new_fungible(DIVISIBILITY_MINIMUM)
                               .flags(MINTABLE)
                               .badge(minter_badge.resource_address(), MAY_MINT)
                               .no_initial_supply(); // Store this resource def on the component
```

Then, inside `add_recipient`, you would mint a new `user_badge` like this: 
```
let user_badge = self.minter_badge_vault.authorize(|auth|{
  self.user_badge_def.mint(1, auth)
});
```